File: twitch/bot.go
package main

import (
	"context"
	"fmt"
	"log/slog"
	"math/rand"
	"strings"
	"time"

	"github.com/nicklaw5/helix/v2"
	twitchws "github.com/vpetrigo/go-twitch-ws"
	"github.com/vpetrigo/go-twitch-ws/pkg/eventsub"
)

type Config struct {
	ClientID          string
	ClientSecret      string
	OAuthToken        string // Streamer's access token (for EventSub subscriptions)
	RefreshToken      string // Streamer's refresh token
	BotOAuthToken     string // Bot's access token (for sending chat messages)
	BotRefreshToken   string // Bot's refresh token
	BotUserID         string
	ChatChannelUserID string
	DbURL             string
	DbAuthToken       string
}

type CommandHandler func(context.Context, *eventsub.ChannelChatMessage) (string, error)
type RewardHandler func(context.Context, *eventsub.ChannelPointsCustomRewardRedemptionAddEvent) error

type Bot struct {
	store          *Store
	helixClient    *helix.Client // Authenticated as streamer (for EventSub)
	botHelixClient *helix.Client // Authenticated as bot (for sending messages)
	cfg            *Config
	rng            *rand.Rand

	commands map[string]CommandHandler
	rewards  map[string]RewardHandler
}

func NewBot(store *Store, cfg *Config) *Bot {
	b := &Bot{
		store:    store,
		cfg:      cfg,
		rng:      rand.New(rand.NewSource(time.Now().UnixNano())),
		commands: make(map[string]CommandHandler),
		rewards:  make(map[string]RewardHandler),
	}

	b.commands = map[string]CommandHandler{
		"!stats":   b.onStatsCommand,
		"!addstat": b.onModifyStatCommand,
		"!rmstat":  b.onModifyStatCommand,
	}

	b.rewards = map[string]RewardHandler{
		"Drink a Potion": b.onDrinkPotion,
		"Tempt the Dice": b.onTemptDice,
	}

	return b
}

func (b *Bot) Connect(ctx context.Context, websocketUrl string) error {
	client := twitchws.NewClient(
		websocketUrl,
		twitchws.WithOnWelcome(b.onWelcomeEvent),
		twitchws.WithOnNotification(b.onNotificationEvent),
		twitchws.WithOnConnect(func() { slog.Info("connected to Twitch websocket") }),
		twitchws.WithOnDisconnect(func() { slog.Warn("disconnected from Twitch websocket") }),
		twitchws.WithOnRevocation(func(_ *twitchws.Metadata, payload *twitchws.Payload) {
			slog.Error("EventSub subscription revoked", "payload", payload)
		}),
		twitchws.WithOnReconnect(func(metadata *twitchws.Metadata, payload *twitchws.Payload) {
			slog.Warn("websocket reconnecting", "metadata", metadata, "payload", payload)
		}))

	if err := client.Connect(); err != nil {
		return fmt.Errorf("connect: %w", err)
	}

	waitErrCh := make(chan error, 1)
	go func() {
		waitErrCh <- client.Wait()
	}()

	select {
	case <-ctx.Done():
		_ = client.Close()
		return ctx.Err()
	case err := <-waitErrCh:
		if err != nil {
			slog.Error("wait error", "err", err)
		}
		_ = client.Close()
		return err
	}
}

func (b *Bot) onWelcomeEvent(_ *twitchws.Metadata, payload *twitchws.Payload) {
	session, _ := payload.Payload.(twitchws.Session)
	slog.Info("received welcome event", "session_id", session.ID)

	ctx := context.Background()

	if err := b.store.InitTokenSchema(ctx); err != nil {
		slog.Error("init token schema failed", "error", err)
		return
	}

	streamerAccessToken := b.cfg.OAuthToken
	streamerRefreshToken := b.cfg.RefreshToken
	if tokens, err := b.store.GetTokens(ctx, "streamer"); err == nil && tokens != nil {
		slog.Info("loaded streamer tokens from database")
		streamerAccessToken = tokens.AccessToken
		streamerRefreshToken = tokens.RefreshToken
	} else {
		slog.Info("using streamer tokens from environment")
	}

	helixClient, err := helix.NewClient(&helix.Options{
		ClientID:        b.cfg.ClientID,
		ClientSecret:    b.cfg.ClientSecret,
		UserAccessToken: streamerAccessToken,
		RefreshToken:    streamerRefreshToken,
	})
	if err != nil {
		slog.Error("create streamer helix client failed", "error", err)
		return
	}

	helixClient.OnUserAccessTokenRefreshed(func(newAccessToken, newRefreshToken string) {
		slog.Info("streamer tokens auto-refreshed")
		if err := b.store.SaveTokens(context.Background(), "streamer", newAccessToken, newRefreshToken); err != nil {
			slog.Error("failed to save streamer tokens during auto-refresh", "error", err)
		}
	})

	b.helixClient = helixClient
	slog.Info("streamer helix client initialised")

	botAccessToken := b.cfg.BotOAuthToken
	botRefreshToken := b.cfg.BotRefreshToken
	if tokens, err := b.store.GetTokens(ctx, "bot"); err == nil && tokens != nil {
		slog.Info("loaded bot tokens from database")
		botAccessToken = tokens.AccessToken
		botRefreshToken = tokens.RefreshToken
	} else {
		slog.Info("using bot tokens from environment")
	}

	botHelixClient, err := helix.NewClient(&helix.Options{
		ClientID:        b.cfg.ClientID,
		ClientSecret:    b.cfg.ClientSecret,
		UserAccessToken: botAccessToken,
		RefreshToken:    botRefreshToken,
	})
	if err != nil {
		slog.Error("create bot helix client failed", "error", err)
		return
	}

	botHelixClient.OnUserAccessTokenRefreshed(func(newAccessToken, newRefreshToken string) {
		slog.Info("bot tokens auto-refreshed")
		if err := b.store.SaveTokens(context.Background(), "bot", newAccessToken, newRefreshToken); err != nil {
			slog.Error("failed to save bot tokens during auto-refresh", "error", err)
		}
	})

	b.botHelixClient = botHelixClient
	slog.Info("bot helix client initialised")

	transport := helix.EventSubTransport{Method: "websocket", SessionID: session.ID}
	subs := []helix.EventSubSubscription{
		{
			Type:      helix.EventSubTypeChannelChatMessage,
			Version:   "1",
			Condition: helix.EventSubCondition{BroadcasterUserID: b.cfg.ChatChannelUserID, UserID: b.cfg.ChatChannelUserID},
			Transport: transport,
		},
		{
			Type:      helix.EventSubTypeChannelPointsCustomRewardRedemptionAdd,
			Version:   "1",
			Condition: helix.EventSubCondition{BroadcasterUserID: b.cfg.ChatChannelUserID},
			Transport: transport,
		},
	}

	for _, sub := range subs {
		resp, err := helixClient.CreateEventSubSubscription(&sub)
		if err != nil {
			slog.Error("eventsub subscription failed", "type", sub.Type, "error", err)
			continue
		}
		slog.Debug("eventsub subscribed", "type", sub.Type, "status", resp.StatusCode)
	}
}

func (b *Bot) onNotificationEvent(_ *twitchws.Metadata, payload *twitchws.Payload) {
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	notification := payload.Payload.(twitchws.Notification)

	switch event := notification.Event.(type) {
	case *eventsub.ChannelChatMessage:
		b.onChatMessage(ctx, event)
	case *eventsub.ChannelPointsCustomRewardRedemptionAddEvent:
		b.onRewardRedemption(ctx, event)
	}
}

func (b *Bot) onChatMessage(ctx context.Context, event *eventsub.ChannelChatMessage) {
	if event.Reply != nil {
		return
	}

	text := strings.TrimSpace(event.Message.Text)
	if len(text) == 0 || text[0] != '!' {
		return
	}

	cmd, _, _ := strings.Cut(strings.ToLower(text), " ")
	handler, ok := b.commands[cmd]
	if !ok {
		return
	}

	response, err := handler(ctx, event)
	if err != nil {
		slog.Error("command error", "command", cmd, "error", err)
		return
	}

	if err := b.send(response, event.MessageID); err != nil {
		slog.Error("command reply failed", "command", cmd, "error", err)
	} else {
		slog.Info("command handled", "command", cmd, "user", event.ChatterUserLogin, "reply", response)
	}
}

func (b *Bot) onRewardRedemption(ctx context.Context, event *eventsub.ChannelPointsCustomRewardRedemptionAddEvent) {
	handler, ok := b.rewards[event.Reward.Title]
	if !ok {
		return
	}

	if err := handler(ctx, event); err != nil {
		slog.Error("reward handler error", "reward", event.Reward.Title, "error", err, "user", event.UserLogin)
	} else {
		slog.Info("reward handled", "reward", event.Reward.Title, "user", event.UserLogin, "cost", event.Reward.Cost)
	}
}

func (b *Bot) send(message, replyToMessageID string) error {
	if b.botHelixClient == nil {
		return fmt.Errorf("bot helix client not initialised")
	}

	params := &helix.SendChatMessageParams{
		BroadcasterID: b.cfg.ChatChannelUserID,
		Message:       message,
		SenderID:      b.cfg.BotUserID,
	}

	if replyToMessageID != "" {
		params.ReplyParentMessageID = replyToMessageID
	}

	_, err := b.botHelixClient.SendChatMessage(params)

	if err != nil && strings.Contains(err.Error(), "Failed to decode API response") {
		time.Sleep(50 * time.Millisecond)
		_, err = b.botHelixClient.SendChatMessage(params)
	}

	return err
}

func (b *Bot) getStatsMessage(ctx context.Context, userID, username string) (string, error) {
	st, err := b.store.GetStats(ctx, userID, username)
	if err != nil {
		return "", err
	}
	return FormatStats(username, st), nil
}

func (b *Bot) randomStat() Stat {
	return statList[b.rng.Intn(len(statList))]
}

func (b *Bot) randomDelta() int {
	if b.rng.Intn(20) == 0 {
		return -1
	}
	return 1
}

File: twitch/src/bot.ts
import { AccessTokenWithUserId, RefreshingAuthProvider } from "@twurple/auth";
import { ApiClient } from "@twurple/api";
import { EventSubWsConfig, EventSubWsListener } from "@twurple/eventsub-ws";
import { Config } from "./config";
import { Store, formatStats, statList, Stat } from "./store";
import { parseModifyStatCommand } from "./command-parser";
import type { EventSubChannelChatMessageEvent } from "@twurple/eventsub-base";
import { log } from "./logger";
import { MockEventSubListener } from "./mock-eventsub";

export class Bot {
  private api: ApiClient;
  private listener: EventSubWsListener;
  private mockListener?: MockEventSubListener;
  private store: Store;
  private authProvider: RefreshingAuthProvider;

  constructor(private config: Config, store: Store) {
    this.store = store;

    this.authProvider = new RefreshingAuthProvider({
      clientId: config.clientId,
      clientSecret: config.clientSecret,
    });

    this.authProvider.onRefresh(async (userId, tokenData) => {
      if (!tokenData.refreshToken) return;

      if (userId === this.config.channelUserId) {
        await this.store.saveTokens(
          "streamer",
          tokenData.accessToken,
          tokenData.refreshToken
        );
        log.info({ userId }, "streamer tokens refreshed");
      }

      if (userId === this.config.botUserId) {
        await this.store.saveTokens(
          "bot",
          tokenData.accessToken,
          tokenData.refreshToken
        );
        log.info({ userId }, "bot tokens refreshed");
      }
    });

    this.authProvider.addUserForToken({
      accessToken: config.streamerAccessToken,
      refreshToken: config.streamerRefreshToken,
      expiresIn: 0,
      obtainmentTimestamp: 0,
    });

    if (config.botUserId !== config.channelUserId) {
      this.authProvider.addUserForToken({
        accessToken: config.botAccessToken,
        refreshToken: config.botRefreshToken,
        expiresIn: 0,
        obtainmentTimestamp: 0,
      });
    }

    this.api = new ApiClient({ authProvider: this.authProvider });
    this.listener = new EventSubWsListener({ apiClient: this.api });
  }

  async init() {
    await this.store.init();
    log.info("store ready");

    await this.refreshTokens();
    log.info("tokens initialised");

    if (this.config.useMockServer) {
      this.mockListener = new MockEventSubListener({
        url: "ws://127.0.0.1:8080/ws",
        onRedemption: async (data) => {
          log.info(
            {
              reward: data.rewardTitle,
              user: data.userName,
              userId: data.userId,
            },
            "channel point reward redeemed"
          );

          if (data.rewardTitle === "Drink a Potion") {
            await this.onDrinkPotion(data.userId, data.userName);
          } else if (data.rewardTitle === "Tempt the Dice") {
            await this.onTemptDice(data.userId, data.userName);
          }
        },
      });
      this.mockListener.start();
      log.info("mock eventsub started");
    } else {
      this.listener.onChannelChatMessage(
        this.config.channelUserId,
        this.config.channelUserId,
        (e) => {
          if (!e.messageText.startsWith("!")) return;

          const [cmd] = e.messageText.toLowerCase().split(" ");
          log.info(
            {
              command: cmd,
              user: e.chatterDisplayName,
              message: e.messageText,
            },
            "chat command received"
          );

          switch (cmd) {
            case "!stats":
              this.handleStats(e.chatterId, e.chatterDisplayName);
              break;
            case "!addstat":
              this.handleModifyStat(e.messageText, e, false);
              break;
            case "!rmstat":
              this.handleModifyStat(e.messageText, e, true);
              break;
          }
        }
      );

      this.listener.onChannelRedemptionAdd(
        this.config.channelUserId,
        async (e) => {
          log.info(
            { reward: e.rewardTitle, user: e.userName, userId: e.userId },
            "channel point reward redeemed"
          );

          if (e.rewardTitle === "Drink a Potion") {
            await this.onDrinkPotion(e.userId, e.userName);
          } else if (e.rewardTitle === "Tempt the Dice") {
            await this.onTemptDice(e.userId, e.userName);
          }
        }
      );

      try {
        this.listener.start();
        log.info("eventsub started");
      } catch (err) {
        log.error({ err }, "eventsub start failed");
      }
    }

    log.info("bot ready");
  }

  private async refreshTokens() {
    const userIds = [this.config.channelUserId];
    if (this.config.botUserId !== this.config.channelUserId) {
      userIds.push(this.config.botUserId);
    }

    const promises = userIds.map((userId) => {
      return new Promise<void>((resolve) => {
        const checkToken = async () => {
          const token = await this.authProvider.getAccessTokenForUser(userId);
          if (token) {
            resolve();
          } else {
            setTimeout(checkToken, 100);
          }
        };
        checkToken();
      });
    });

    await Promise.all(promises);
  }

  private async sendMessage(message: string) {
    await this.api.asUser(this.config.botUserId, async (ctx) => {
      await ctx.chat.sendChatMessage(this.config.channelUserId, message);
    });
    log.debug({ message }, "message sent");
  }

  private async handleStats(userId: string, username: string) {
    const stats = await this.store.getStats(userId, username);
    await this.sendMessage(formatStats(username, stats));
    log.info({ userId, username }, "stats command handled");
  }

  private async handleModifyStat(
    command: string,
    event: EventSubChannelChatMessageEvent,
    isRemove: boolean
  ) {
    const isMod = event.badges.moderator || event.badges.broadcaster;
    if (!isMod) {
      log.warn(
        { user: event.chatterDisplayName, command },
        "non-moderator attempted to use mod command"
      );
      await this.sendMessage("You must be a moderator to use this command");
      return;
    }

    const parsed = parseModifyStatCommand(command, isRemove);
    if (parsed.error) {
      log.warn({ command, error: parsed.error }, "invalid modify stat command");
      await this.sendMessage(parsed.error);
      return;
    }

    const { mentionedLogin, statColumn, amount } = parsed;

    await this.store.modifyStat(
      event.chatterId,
      mentionedLogin,
      statColumn,
      isRemove ? -amount : amount
    );

    const stats = await this.store.getStats(event.chatterId, mentionedLogin);
    await this.sendMessage(formatStats(mentionedLogin, stats));
    log.info(
      {
        moderator: event.chatterDisplayName,
        target: mentionedLogin,
        stat: statColumn,
        amount: isRemove ? -amount : amount,
      },
      "modify stat command handled"
    );
  }

  private randStat(): Stat {
    return statList[Math.floor(Math.random() * statList.length)];
  }

  private randDelta(): number {
    return Math.random() < 0.05 ? -1 : 1;
  }

  private async onDrinkPotion(userId: string, userName: string) {
    const stat = this.randStat();
    const delta = this.randDelta();
    const outcome = delta < 0 ? "lost" : "gained";

    await this.store.modifyStat(userId, userName, stat.column, delta);

    await this.sendMessage(
      `A shifty looking merchant hands ${userName} a glittering potion. Without hesitation, they sink the whole drink. ${userName} ${outcome} ${stat.display}`
    );

    const stats = await this.store.getStats(userId, userName);
    await this.sendMessage(formatStats(userName, stats));

    log.info(
      { userId, userName, stat: stat.column, delta, outcome },
      "drink potion reward handled"
    );
  }

  private async onTemptDice(userId: string, userName: string) {
    await this.sendMessage(`${userName} has rolled with initiative.`);

    const stats = await this.store.getStats(userId, userName);
    await this.sendMessage(formatStats(userName, stats));

    log.info({ userId, userName }, "tempt dice reward handled");
  }
}
